src/core/file.cc

```diff
@@ -1048,20 +1082,60 @@ static internal::alignments xfs_alignments(const dioattr& da, unsigned block_siz
     return ret;
 }
 
+// Query block device alignment properties using ioctl and statx
+static
+internal::alignments
+blkdev_alignments(int fd, dev_t device_id) {
+    // Query logical block size (smallest addressable unit)
+    size_t logical_block_size;
+    if (auto ret = ::ioctl(fd, BLKBSZGET, &logical_block_size); ret == -1) {
```

Comment:
Is BLKBSZGET not intended for the filesystem block size?

See comments to this response: https://unix.stackexchange.com/a/52225/343739

Reply:
Thanks for pointing this out, Avi. I am quoting the comments here:

> Well, I realized that [block/ioctl.c](http://lxr.free-electrons.com/source/block/ioctl.c) in Linux kernel source contains misleading comments. BLKPBSZGET (notice the P) gets physical sector size, BLKSSZGET gets logical sector size and BLKBSZGET (or BLKBSZGET_32 in block/compat_ioctl.c) gets file system allocation unit (cluster) size. [Logical and physical sector size differ when Advanced Format 512e is used.](http://www.oeey.com/2014/10/invalid-argument-directio-4kn.html) ([512e @ WP](http://en.wikipedia.org/wiki/Advanced_Format#512e)) – 
> [Palec](https://unix.stackexchange.com/users/50602/palec)

so, i am switching to  `BLKSSZGET` for the `logical_block_size` in `blkdev_alignments()`.